csds-client/client/client.go

```diff
@@ -0,0 +1,188 @@
+	CsdsClient csdspb.ClientStatusDiscoveryServiceClient
+
+	Nm   []*envoy_type_matcher.NodeMatcher
+	Md   metadata.MD
+	Info Flag
+}
+
+// parse flags to info
+func ParseFlags() Flag {
+	uriPtr := flag.String("service_uri", "trafficdirector.googleapis.com:443", "the uri of the service to connect to")
+	platformPtr := flag.String("cloud_platform", "gcp", "the cloud platform (e.g. gcp, aws,  ...)")
+	authnModePtr := flag.String("authn_mode", "auto", "the method to use for authentication (e.g. auto, jwt, ...)")
+	apiVersionPtr := flag.String("api_version", "v2", "which xds api major version  to use (e.g. v2, v3 ...)")
+	requestFilePtr := flag.String("request_file", "", "yaml file that defines the csds request")
+	requestYamlPtr := flag.String("request_yaml", "", "yaml string that defines the csds request")
+	jwtPtr := flag.String("jwt_file", "", "path of the -jwt_file")
+	configFilePtr := flag.String("file_to_save_config", "", "the file name to save config")
+
+	flag.Parse()
+
+	f := Flag{
+		uri:         *uriPtr,
+		platform:    *platformPtr,
+		authnMode:   *authnModePtr,
+		apiVersion:  *apiVersionPtr,
+		RequestFile: *requestFilePtr,
+		RequestYaml: *requestYamlPtr,
+		jwt:         *jwtPtr,
+		configFile:  *configFilePtr,
+	}
+
+	return f
+}
+
+// parse the csds request yaml to nodematcher
+func (c *Client) ParseNodeMatcher() error {
+	if c.Info.RequestFile == "" && c.Info.RequestYaml == "" {
+		return fmt.Errorf("missing request yaml")
+	}
+
+	var nodematchers []*envoy_type_matcher.NodeMatcher
+	err := ParseYaml(c.Info.RequestFile, c.Info.RequestYaml, &nodematchers)
+	if err != nil {
+		return fmt.Errorf("%v", err)
+	}
+
+	c.Nm = nodematchers
+	return nil
+}
+
+// connect uri with authentication
+func (c *Client) ConnWithAuth() error {
+	var scope string
+	if c.Info.authnMode == "jwt" {
+		if c.Info.jwt == "" {
+			return fmt.Errorf("missing jwt file")
+		} else {
+			if c.Info.platform == "gcp" {
+				scope = "https://www.googleapis.com/auth/cloud-platform"
+				pool, err := x509.SystemCertPool()
+				creds := credentials.NewClientTLSFromCert(pool, "")
+				perRPC, err := oauth.NewServiceAccountFromFile(c.Info.jwt, scope)
+				if err != nil {
+					return fmt.Errorf("%v", err)
+				}
+
+				c.Cc, err = grpc.Dial(c.Info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+				if err != nil {
+					return fmt.Errorf("%v", err)
+				} else {
+					return nil
+				}
+			}
+		}
+	} else if c.Info.authnMode == "auto" {
+		if c.Info.platform == "gcp" {
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewApplicationDefault(context.Background(), scope) // Application Default Credentials (ADC)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+
+			// parse GCP project number as header for authentication
+			if projectNum := ParseGCPProject(c.Nm); projectNum != "" {
```

Comment:
should probably error out if no project number?

Reply:
Since project number is only required for auto authentication mode on local machine running with gCloud, for GCE and GKE there's no need to add project number to the header for authentication, I returned empty string if no project number here.
However, for GCP, project number is required in request yaml, but maybe this should be field missing error of passing request yaml but not be stated here? What do you think?

Reply:
Yes, we can error out when parsing input. For GCP, both project number and network name are required.

Reply:
Added check in `parseNodeMatcher()`. The required keys are specified based on platform and uri.